Full market-simulation

sygi opened the pull request:
Introduces `HybridMarketSimulator`, which first executes like `FxMarketSimulator` (but actually deals with the `LimitOrders`...), and after getting a `'ChangeMarketRules` message changes execution to the regular simulation (like `OrderBookMarketSimulator`. To do it, I added a class of `MarketRulesWrapper` - the _really_ self-contained MarketRules.

sygi committed:
market simulation with historical data as seed. Implements creating quotes in a dummy way

mogli91 committed:
[Fund Distributions]
added file FundDistributions.scala
Each country is implemented as a class that inherits from FundDistribution.
Statistics are entered in tuples (amount in local currency, percentage of people earning this amount).
Internally, a list of integers is created that contains as many entries of
a particular amount as the percentage indicates.
When calling "get" a sample out of the internal list is returned using a Random object.

sygi committed:
LimitOrders for ForexMarketRules, wrapper around MarketRules, HybridMarketSimulator to change the policy during the execution

sygi committed:
Merge branch 'full-market' of http://github.com/merlinND/TradingSimulation into full-market

merlinND commented on ts/src/main/scala/ch/epfl/ts/brokers/StandardBroker.scala:
```diff
@@ -80,6 +80,7 @@ class StandardBroker extends Component with ActorLogging {
       })
     }
 
+    //TODO(sygi): refactor
```
Would you like to include that refactoring in this PR?

sygi replied:
Generally done, it would be best to actually have super class for ExecutedOrders, but something wasn't working there.

merlinND commented on ts/src/main/scala/ch/epfl/ts/brokers/StandardBroker.scala:
```diff
@@ -110,6 +111,11 @@ class StandardBroker extends Component with ActorLogging {
       val replyTo = sender
       val uid = o.chargedTraderId()
+      if (!ableToProceed(o)){
+        log.debug("Broker: Unable to proceed MarketBid request before getting first quote")
```
Maybe another log level ("info" or "warning"?) would be more appropriate in this case

sygi replied:
done

merlinND commented on ts/src/main/scala/ch/epfl/ts/component/utils/BackLoop.scala:
```diff
@@ -14,10 +19,16 @@ class BackLoop(marketId: Long, p: Persistance[Transaction]) extends Component {
+
+  def createQuote(t: Transaction){
+    return Quote(t.mid, t.timestamp, t.whatC, t.withC, t.price, t.price)
+    //TODO(sygi): to implement it properly, we need working MarketSimulator (the one that keeps orderbooks)
```
Do you think this could be implemented in this PR?

sygi replied:
I actually forgot to remove it, done.

merlinND commented on ts/src/main/scala/ch/epfl/ts/config/FundDistribution.scala:
```diff
@@ -0,0 +1,70 @@
+      (375000, 0.4),
+      (500000, 0.1),
+      (1000000, 0.1),
+      (1000001, 0.1));
```
Why make a difference between `1000000` and `1000001`?

merlinND commented on ts/src/main/scala/ch/epfl/ts/config/FundDistribution.scala:
```diff
@@ -0,0 +1,70 @@
+  var bins : List[Int];
+  val rand = new java.util.Random();
+  
+  def percentListToDistribution(binsInPercent: List[(Int, Double)]) : List[Int] = {
```
Add an assert which checks that the distribution being passed sums up to 100%

merlinND replied:
Create a test suite that checks this function's output on a few simple distributions.

merlinND commented on ts/src/main/scala/ch/epfl/ts/config/FundDistribution.scala:
```diff
@@ -0,0 +1,70 @@
+ */
+
+trait FundDistribution {
+  var bins : List[Int];
```
Make it a `def`: subclasses will still be able to override it, and it makes more sense since the bins actually do not vary.

merlinND commented on ts/src/main/scala/ch/epfl/ts/config/FundDistribution.scala:
```diff
@@ -0,0 +1,70 @@
+      (6000, 20.9),
+      (7000, 12.9),
+      (8000, 8.3),
+      (8001, 19.3)).map(x => (x._1 * 12, x._2));
```
What is this `* 12` magic number?

merlinND commented on ts/src/main/scala/ch/epfl/ts/engine/HybridMarketSimulator.scala:
```diff
@@ -0,0 +1,38 @@
+ * Created by sygi on 09.05.15.
+ */
+class HybridMarketSimulator(marketId: Long, rules1: FxMarketRulesWrapper, rules2: MarketRulesWrapper) extends MarketSimulator(marketId, rules1.getRules) {
+  var rulesChanged = false
```
![The rules](https://media.licdn.com/mpr/mpr/p/1/005/06a/14d/28dda2d.jpg)

merlinND commented on ts/src/main/scala/ch/epfl/ts/engine/HybridMarketSimulator.scala:
```diff
@@ -0,0 +1,38 @@
+    case o: Order =>
+      getCurrentRules.processOrder(o, marketId, book, tradingPrices, this.send[Streamable])
+    case 'ChangeMarketRules =>
+      println("Hybrid market: changing rules")
```
Change to `log.info`

sygi replied:
I can't - I don't have access to `context` here.

merlinND commented on ts/src/main/scala/ch/epfl/ts/engine/MarketFXSimulator.scala:
```diff
@@ -15,7 +15,7 @@ class MarketFXSimulator(marketId: Long, rules: ForexMarketRules) extends MarketS
 
     case limitAsk: LimitAskOrder =>
-    // TODO
+    // TODO - add limit order resolvation based on given quotes
```
Could it be included in this PR?

Actually, wasn't it implemented in the previous project when they were focusing on Bitcoin?

sygi replied:
In `FullMarketSimulation` I don't use `MarketFXSimulator`. But I implemented something that could be put here in the next PR in `ForexMarketRules`.

sygi replied:
And no, because they don't care about `Quote`s at all and only keep order books. If we use their strategy, we'll get two divergent symbol prices, one for `LimitOrder`s and one for `MarketOrder`s.

merlinND commented on ts/src/main/scala/ch/epfl/ts/engine/MarketRules.scala:
```diff
@@ -111,14 +117,25 @@ class MarketRules extends Serializable {
+      println("MR: generating quote " + topBid.price + " " + topAsk.price)
+      send(Quote(marketId, timestamp, topAsk.whatC, topAsk.withC, topBid.price, topAsk.price))
+    } else
+      println("MR: can't generate quote")
```
Should it be a more serious error?

sygi replied:
1. this is the expected behavior
2. it can't be (at least I don't see how to do it) - see the comment about changing to `log.info` above.

merlinND commented on ts/src/main/scala/ch/epfl/ts/engine/rules/FxMarketRulesWrapper.scala:
```diff
@@ -0,0 +1,90 @@
+ * Created by sygi on 09.05.15.
+ */
+class FxMarketRulesWrapper extends MarketRulesWrapper(new ForexMarketRules()) {
+  val rules = new ForexMarketRules()
```
I see two `new ForexMarketRules`, couldn't you only use a `val` constructor parameter in the `MarketRulesWrapper`?

sygi replied:
I tried, but I don't know how to do it :(

sygi replied:
Ok, now I know, done.

merlinND commented on ts/src/main/scala/ch/epfl/ts/engine/rules/FxMarketRulesWrapper.scala:
```diff
@@ -0,0 +1,90 @@
+      book.bids.insert(limitBid)
+      checkPendingOrders(marketId, book, tradingPrices, send)
+
+    case limitAsk: LimitAskOrder =>
```
Would it be possible to merge these two cases with `|`?

sygi replied:
They don't do the same...

merlinND replied:
Oh, my bad, I read too fast

merlinND commented on ts/src/main/scala/ch/epfl/ts/engine/rules/FxMarketRulesWrapper.scala:
```diff
@@ -0,0 +1,90 @@
+      tradingPrices.get((marketAsk.withC, marketAsk.whatC)) match {
+        // We sell at current bid price
+        case Some(t) => rules.matchingFunction(marketId, marketAsk, book.bids, book.asks, send, t._1)
+        //TODO(sygi): does it actually work at all, when there are multiple currencies?
```
Would it be possible to verify that with a test suite? (If it's not too long to write)

sygi replied:
I thought about it and it actually don't work, as there is only one instance of `OrderBook`, so one cannot take the top bid/ask from the given currency.

merlinND commented on ts/src/main/scala/ch/epfl/ts/engine/rules/FxMarketRulesWrapper.scala:
```diff
@@ -0,0 +1,90 @@
+      println("FMRW: got unknown order")
+  }
+
+  //TODO(sygi): refactor common code
```
Try including the refactoring in this PR if possible.

sygi replied:
I am actually not sure if this will be productive, as they differ in few things and I would have to put `if`s instead.

merlinND commented on ts/src/main/scala/ch/epfl/ts/engine/rules/MarketRulesWrapper.scala:
```diff
@@ -0,0 +1,18 @@
+/**
+ * A wrapper around the MarketRules, so that all the information needed to proceed with the order is in one place.
+ * Needed to be able to change between strategies without changing the MarketSimulator.
+ * Created by sygi on 09.05.15.
```
Remove these

sygi replied:
You meant created by comments? If so, done.

merlinND commented on ts/src/main/scala/ch/epfl/ts/engine/rules/SimulationMarketRulesWrapper.scala:
```diff
@@ -0,0 +1,69 @@
+ * Created by sygi on 09.05.15.
+ */
+class SimulationMarketRulesWrapper() extends MarketRulesWrapper(new MarketRules()){
+  val rules = new MarketRules()
```
Same comment about re-using `rules`

sygi replied:
done.

merlinND commented on ts/src/main/scala/ch/epfl/ts/example/FullMarketSimulation.scala:
```diff
@@ -0,0 +1,148 @@
+ */
+object FullMarketSimulation {
+  var producers = Map[Class[_], List[ComponentRef]]()
+  var consuments = Map[Class[_], List[ComponentRef]]()
```
Oh, [okay](http://en.wiktionary.org/wiki/consument).

sygi replied:
It's actually a popular [terminology](http://en.wikipedia.org/wiki/Producer%E2%80%93consumer_problem)

merlinND commented on ts/src/main/scala/ch/epfl/ts/example/FullMarketSimulation.scala:
```diff
@@ -0,0 +1,148 @@
+  def addEntity(msg: Class[_], component: ComponentRef, mapping: Map[Class[_], List[ComponentRef]]) =
+    mapping + (msg -> (component :: mapping.getOrElse(msg, List())))
+
+  def connectProducersWithConsuments() = {
```
All these functions are really nice but unrelated to full market simulation, could they be refactored to a superclass and used in all our example classes?

sygi replied:
It can be done (in the next PR), but actually I don't see the gain from using it when we have only one creator of `Quotes` or one consument of it.

merlinND commented on ts/src/main/scala/ch/epfl/ts/example/FullMarketSimulation.scala:
```diff
@@ -0,0 +1,148 @@
+  def connectAllOrders(source: ComponentRef, destination: ComponentRef) =
+    source->(destination, classOf[LimitBidOrder], classOf[LimitAskOrder], classOf[MarketBidOrder], classOf[MarketAskOrder])
+
+  def createFetcher(useLiveData: Boolean, builder: ComponentBuilder, symbol: (Currency, Currency)) = {
```
Same comment, this would be useful in most of our other examples

merlinND replied:
You can make `builder` an `implicit` parameter.

sygi replied:
I don't like implicit parameters.

merlinND commented on ts/src/main/scala/ch/epfl/ts/example/FullMarketSimulation.scala:
```diff
@@ -0,0 +1,148 @@
+        market.ar ! 'ChangeMarketRules
+      }
+    }
+    timer.schedule(new StopFetcher, 10 * 1000)
```
Take the delay as a parameter

sygi replied:
done.

merlinND commented on ts/src/main/scala/ch/epfl/ts/engine/MarketRules.scala:
```diff
@@ -47,13 +53,13 @@ class MarketRules extends Serializable {
                        matchExists: (Double, Double) => Boolean = alwaysTrue,
                        oldTradingPrice: Double,
                        enqueueOrElse: (Order, PartialOrderBook) => Unit): Double = {
```
Could this function be refactored? Right now it's very difficult to read and understand what it does.

sygi replied:
I hardly know what it does, it has been here since we started and I believe that if I tried to change something here, everything will break.

mogli91 committed:
[CONFIG] added distributions for all currencies

mogli91 committed:
added assertion to check whether percentages add up to 100.
needed to change some values to fulfill this requirement (rounding errors screw up the 100%)

mogli91 committed:
added random assignments for entries "earning more than xxx" (it is computed as value * (1 + randomdouble)

mogli91 committed:
*removed semicolons at end of line
*added basic test suite for FundDistribution

mogli91 committed:
added some more tests for FundDistribution

sygi committed:
applying review comments (#45)

sygi committed:
Merge branch 'full-market' of http://github.com/merlinND/TradingSimulation into full-market

sygi committed:
removing auto generated comments

sygi committed:
slower speed of fetcher

sygi committed:
default constructor values in MarketRulesWrappers

sygi committed:
documenting code

sygi committed:
correcting the tests

sygi committed:
Merge branch 'master' into full-market

Conflicts:
	ts/src/main/scala/ch/epfl/ts/component/Component.scala
	ts/src/main/scala/ch/epfl/ts/component/fetch/HistDataCSVFetcher.scala
	ts/src/main/scala/ch/epfl/ts/example/MovingAverageFXExample.scala
	ts/src/main/scala/ch/epfl/ts/traders/MadTrader.scala
	ts/src/test/scala/ch/epfl/ts/test/TestHelpers.scala

sygi committed:
instantiating a marketFx with MarketRulesWrapper

sygi commented:
![screenshot](https://cloud.githubusercontent.com/assets/495380/7571359/64e309ae-f817-11e4-84d4-0ecc624b7b31.png)
let's :shipit:

sygi committed:
dummy messgae

merlinND committed:
Introduce `Timekeeper` as a way to keep track of simulated time

merlinND committed:
Simple associated test suite

merlinND committed:
Use `Timekeeper` when simulating to keep consistent time scale

merlinND committed:
Update test suite to make it less random

merlinND committed:
Empty `receiver` function is actually enough

merlinND committed:
Merge pull request #51 from merlinND/add-gps

Add fancy GPS system

merlinND commented:
Let's just wait on Travis and we should be able to merge this one too.

sygi commented:
:shipit:
